prov/tcp: Move progress mgmt to domain

shefty opened the pull request:
Introduce a new progress engine.  The engine/class encapsulate progress
management into a single object.  The engine will allow us to move progress from the
current CQ + counters to a single location.  This will not only simplify the code, but
make it easier to apply future enhancements, such as support for io_uring.  The
progress engine is associated with an open fabric object.

All tcp progress is migrated to the new progress engine.  This includes all connection
management, listening sockets, connecting sockets, and active endpoints.  As a result of
these changes, the progress locking is restructured and simplified.  There's no longer any
need to bounce a socket between different poll sets (EQ and CQ, for example), or add the
socket to multiple poll sets (e.g. Tx and Rx CQs).  Connections should
also be established quicker, since listening sockets are watched even when trying to drive
data progress.  (A future patch will improve this more by allowing tcp to callback into
rxm for faster connection setup).

The expectation is that applications that do not use wait object or blocking calls should
see a slight improvement to stability and performance.  For applications that use wait
objects or blocking calls, stability improvements are expected, but the impact on
performance is unknown as of now.  Blocking calls require spawning a progress thread
to drive progress.  It may be possible to remove the progress thread in certain cases
as a future enhancement, but for now, the thread helps ensure correct behavior prior
to implementing other enhancements.

shefty commented:
@sydidelot - Here's the change to move progress to the domain.  Only change from before is rebasing the code to the latest main and resolving the conflicts.  Compiles, but otherwise untested, as it needs a lot more work.

sydidelot commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -846,3 +813,159 @@ static ssize_t (*tcpx_start_op[ofi_op_write + 1])(struct tcpx_ep *ep) = {
+		(void) tcpx_update_epoll(ep);
+		ofi_mutex_unlock(&ep->lock);
+	}
+	ofi_mutex_unlock(&domain->progress_lock);
```
I found an interesting bug while testing the patch: the function keeps accessing an endpoint after it is closed.
The problem is that `domain->progress_lock` is released early in the function, then the function calls `ofi_{epoll/pollfds}_wait()` and processes ready events. If the endpoint is closed after the call to `ofi_{epoll/pollfds}_wait()` but before processing the events, the function may access a dead endpoint (e.g., the log `ep->lock` would already been destroyed).

One way to work around this problem is to keep the lock `domain->progress_lock` until the end of the function (like it was in the former code).

shefty replied:
Accessing the ep after its closed is an area I'm actively concerned about.  I'm not seeing an alternative to holding the progress_lock throughout the call.

sydidelot commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -846,3 +813,159 @@ static ssize_t (*tcpx_start_op[ofi_op_write + 1])(struct tcpx_ep *ep) = {
+	while (domain->auto_progress) {
+		ofi_mutex_unlock(&domain->progress_lock);
+
+		// TODO: make blocking to avoid spinning
```
What strategy do you have in mind to avoid spinning here? I thought about the following code:

```
struct ofi_epollfds_event events[1];

if (domain->wait_obj == FI_WAIT_FD) {
	nfds = ofi_epoll_wait(domain->epoll_fd, events,
			      1, -1);
} else {
	nfds = ofi_pollfds_wait(domain->pollfds, events,
				1, -1);
}

if (nfds > 0)
	tcpx_progress_domain(domain);
```

I think calling `ofi_{epoll/pollfds}_wait()` multiple times without processing the events is ok since the poll set is configured for level-triggered. 
What do you think?

shefty replied:
I'm hoping something like what you have will work.  I'd like to avoid calling poll() multiple times and discarding the events from one of the calls, since poll() appears to be a slow operation when watching across a large number of sockets.  In that case, we'd need to pass a timeout to the progress_domain() call, but figure out a way to avoid use after close.

sydidelot reviewed (changes requested):
Hi @shefty. I have tested the code over the week-end and I found a small set of issues.
Please take a look at my comments. Thanks for the patch BTW!

shefty commented:
@sydidelot - thanks for the comments.  I'll look the comments in relation to the code in detail when I get a chance.  I need to re-examine how wait support can work without always introducing a progress thread.  We have an app that relies on blocking on the fd's.

shefty commented on prov/tcp/src/tcpx_conn_mgr.c:
```diff
@@ -291,10 +192,16 @@ static int tcpx_ep_enable(struct tcpx_ep *ep,
 
+	// TODO: move to create/close - do we need state check in progress loop?
+	// want to continue to process receives as long as there's data
+	ofi_genlock_lock(&domain->util_domain.lock);
```
I think you are correct.  I'll update

shefty commented on prov/tcp/src/tcpx_cq.c:
```diff
@@ -382,15 +321,23 @@ int tcpx_cntr_open(struct fid_domain *fid_domain, struct fi_cntr_attr *attr,
+	ofi_mutex_lock(&domain->progress_lock);
 	ret = ofi_cntr_init(&tcpx_prov, fid_domain, attr, cntr,
 			    &tcpx_cntr_progress, context);
+	ofi_mutex_unlock(&domain->progress_lock);
```
I don't remember why I added the progress_lock here.  Doesn't seem needed.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -545,45 +538,25 @@ static ssize_t tcpx_ep_cancel(fid_t fid, void *context)
-	 * socket from the pollfd set to disable progress.
-	 */
+	tcpx_unwatch_socket(domain, ep);
+	ofi_genlock_lock(&domain->util_domain.lock);
```
thanks - will update

shefty commented:
Updated with some locking changes.  I'm still not sure about the locking around cq and cntr init.  I kept the locks there for now, but still don't see why I added them.  Blocking of the progress thread is still needed.  Before creating a progress thread when blocking cq calls are needed, I need to ensure that it doesn't impact performance.

shefty commented:
@sydidelot - I'm finally getting back to this.  I'm considering moving the progress up to the fabric object, so the EQ can progress connection requests using the same poll fds as the CQs.  That is, listening sockets could be added to the epoll, rather than watched separately.  The intent is to be more responsive to connection establishment.  Would this change impact your application?

shefty commented:
I updated this PR with new patches that should be in decent enough shape to evolve into a mergeable solution.  The first patch defines a new progress engine mechanism, which basically re-uses existing code but contained in a class-like abstraction.  (The existing code that it is based on will be removed in follow on patches, so that there's no duplicated code at the end).  The progress engine is setup to handle progress on all sockets.  The second patch adds passive endpoints/listening sockets to the progress engine.  As of now, the progress engine will only handle listening sockets, so only the tcpx_accept() path will be taken.

My plan is to migrate the 'CM' watched sockets from the EQ to the progress engine, followed by migrating the active endpoints.

shefty commented:
Still debugging CI issues.  I've been able to see a delay while running fi_ubertest (specifically test case 92 from tcp/all).  The problem appears to be an existing issue in the underlying pollfds implementation.  It looks like some sort of race between updating the pollfd set and signaling the progress thread.  I've identified a problem in the call ordering of the code, but haven't found what the fix is yet.

sydidelot commented on prov/tcp/src/tcpx.h:
```diff
@@ -282,10 +285,67 @@ struct tcpx_ep {
+
 struct tcpx_fabric {
 	struct util_fabric	util_fabric;
+	struct tcpx_progress	progress;
```
Say I have a multi-threaded application where each thread runs with its own collection of EPs, EQs and CQs, all belonging to the same fabric (operations accessing the same resources are serialized on the same thread).

What will happen if all the threads concurrently run `fi_cq_read()`? Will they all compete for the same progress lock? If so, what do you suggest for multi-threaded applications that want to run lockless? Do they need to create one fabric object per thread?

shefty replied:
Currently, the progress engine is associated with the fabric object.  As long as 2 CQs fall under the same fabric, they will compete for the progress lock.  As the patches are structured, each thread would need to create its own fabric object to run lockless (eventually).

I looked at associating the progress engine with the domain.  The issue is how to handle listening sockets.  Another scaling issue I've seen is slowness setting up new connections, so I want to be able to monitor new connections for faster setup.  My plan is to follow these patches with a 'fast connect' option that rxm can use.  This would allow tcp to invoke a callback after accepting a new socket, so the rxm CM messages can be sent immediately, without needing to write and process events on the EQ.  (When there's a 1000 active connections all sending and receiving data, it can take the app a long time to get around to checking the EQ to setup a new connection.  And most new connections require that the server process 2 EQ events before the EP is usable.)

There are a couple other options for how to handle listening sockets, but this seemed the easiest approach for now.

shefty commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -853,3 +853,369 @@ static ssize_t (*tcpx_start_op[ofi_op_write + 1])(struct tcpx_ep *ep) = {
+	int nfds, i;
+	bool pin, pout, perr;
+
+	ofi_mutex_lock(&progress->lock);
```
Yes, I would like to allow lockless execution.  MPI can also use that.  There are currently a whole collections of locks in use through the progress path, all of which need to be addressed -- CQ lock, EQ lock, progress lock, EP lock, fd signaling lock, poll set lock, etc.  I'm not even sure if I listed them all.  Toss rxm over this, and there's more.

shefty commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -853,3 +853,369 @@ static ssize_t (*tcpx_start_op[ofi_op_write + 1])(struct tcpx_ep *ep) = {
+
+void tcpx_close_progress(struct tcpx_progress *progress)
+{
+	assert(dlist_empty(&progress->ep_list));
```
Yes, we can do that.

shefty commented on prov/tcp/src/tcpx_eq.c:
```diff
@@ -38,10 +38,13 @@
 	eq = container_of(eq_fid, struct util_eq, eq_fid);
+	fabric = container_of(eq->fabric, struct tcpx_fabric, util_fabric);
 
+	tcpx_run_progress(&fabric->progress);
```
One of the changes is to remove the ep list, so we never iterate over all EPs.  I believe this is partially contributing to scaling issues.  The ep_list became the active_wait_list, which should be a much smaller set of EPs.  But, you are correct that reading the CQ and reading the EQ will both execute progress.

shefty commented:
As a general response, I have 2 specific applications that I'm working with, both of which are having some issues at scale.  MPI is single threaded, but is seeing slow performance at about 500+ ranks for specific applications.  There's a definite problem with scaling, since the performance gap increases (relative to a baseline) as the number of ranks increase.  The problems are limited to specific, but critical, applications.  Some of the changes here, such as replacing the ep_list, eliminating the need to update multiple poll sets (1 per CQ), are to help address this.  The other change is a future enhancement, which is the fast connect option.

The other application is a multi-threaded app that uses blocking calls (directly waiting of the fd's) and has peers coming and going at all times, even during data transfers.  Part of the changes here are to simplify the locking to help improve stability.

Am aware of your app, and I won't merge any changes until I can confirm that these changes do not make performance worse for any of the apps.  But since I'm still working on the changes not breaking the CI, merging isn't immanent.  (At least some of the problems are in existing code that the changes are exposing.)

shefty commented:
I'm updated the changes to associate a progress engine with the fabric (to drive EQ related progress - basically listening sockets) and adding a progress engine to each domain (to drive CQ/data transfer progress).  I'll open a new PR once those changes are ready to fail CI.  :)

sydidelot commented:
@shefty Our application uses 2 different polling models. The server spins for completion and never blocks. As for the client, it retrieves the FDs corresponding to the CQs/EQs, creates an epoll set ond blocks on it.
